### PR TITLE
[ESP8266/ESP32] Problem setting hostname while on dynamic ip.

### PR DESCRIPTION
**Pull request by pcniatic**
While using a dynamic ip the system doesn't take into account the hostname setting.
Calling WiFi.hostname() after begin() solved the issue.
This only applies to esp8266, haven't tried on esp32.

### CODE REVIEW

**Comment by rjwats** [comment]
Both ESP32 and esp8266 will not accept a dhcp hostname until the WiFi is set to an active mode. It looks like the advice is to put the device into the correct mode, and set the hostname before calling WiFi.begin().

Your fix effectively does this, however I'm not able to properly test this approach under ESP32 because there is another issue affecting the DHCP hostname feature at the moment. There is a pending PR to fix the issue:

https://github.com/espressif/arduino-esp32/issues/2537
https://github.com/espressif/esp-lwip/pull/6

I'll keep an eye on this. Meanwhile would you mind moving the hostname calls for both devices below the begin call, because it's clearly wrong in both instances, so we can merge it in.

**Comment by pcniatic** [comment]
This is getting really interesting/fun :-). I have some thoughts/findings regarding this topic:

1) I am not sure, but **does it make sense to set the hostname while on static ip?** From the ESP8266WiFi doc it only applies to dhcp: [esp8266wifi/station-class.html#hostname](https://arduino-esp8266.readthedocs.io/en/latest/esp8266wifi/station-class.html#hostname)

2) After reading your post it kept wonder in my mind the idea that dhcp hostname should be set before WiFi.begin(). Reading the code of ESP8266WiFiSTAClass::config() and ESP8266WiFiSTAClass::begin() the first thing they do is enabling STA with `WiFi.enableSTA(true)`. Although it is not documented, it is at least named as a not properly documented functions: [esp8266wifi/generic-class.html#other-function-calls](https://arduino-esp8266.readthedocs.io/en/latest/esp8266wifi/generic-class.html#other-function-calls).
Pretty much it checks if STA is previously active, otherwise is activated along with whatever is already activated. This is the code:
```
/**
 * control STA mode
 * @param enable bool
 * @return ok
 */
bool ESP8266WiFiGenericClass::enableSTA(bool enable) {

    WiFiMode_t currentMode = getMode();
    bool isEnabled = ((currentMode & WIFI_STA) != 0);

    if(isEnabled != enable) {
        if(enable) {
            return mode((WiFiMode_t)(currentMode | WIFI_STA));
        } else {
            return mode((WiFiMode_t)(currentMode & (~WIFI_STA)));
        }
    } else {
        return true;
    }
}
```
Long story short, I tried calling `WiFi.enableSTA(true)` and then setting the hostname prior to `WiFi.begin()` and it worked!
```
void WiFiSettingsService::reconfigureWiFiConnection() {
    Serial.println("Reconfiguring WiFi...");

    // disconnect and de-configure wifi and software access point
    WiFi.disconnect(true);

    WiFi.enableSTA(true);

    // configure static ip config for station mode (if set)
    if (_staticIPConfig) {
      WiFi.config(_localIP, _gatewayIP,  _subnetMask, _dnsIP1, _dnsIP2);
    }

    // connect to the network
#if defined(ESP8266)
    WiFi.hostname(_hostname);
#elif defined(ESP_PLATFORM)
    WiFi.setHostname(_hostname.c_str());
#endif

    WiFi.begin(_ssid.c_str(), _password.c_str());
}
```

**What do you think?**

**Comment by pcniatic** [comment]
Finally I came up with this. While trying to solve the hostname issue, I found that changing from static ip to dynamic ip was not working. Investing the WiFi API, I found there is a variable named `_useStaticIp` that is only being set inside WiFi.config. So, I added a new condition in WiFiSettingService.cpp, and now when _staticIPConfig is false I also call WiFi.config but with a INADDR_ANY (esp8266) / INADDR_NONE (esp32) value. This solved the issue of moving from static to dynamic ip on both platforms, AND IT ALSO SOLVED THE HOSTNAME ISSUE!!
Let me know what you think.
Best, Mauro.

**Comment by rjwats** [comment]
Yes, the hostname is only used during the DHCP request - so it is not
needed if using static IP.

Looks like your solution of calling the config function has the side effect
of enabling STA mode and restarting DHCP - this works around the issue on
the ESP32. Very good!

I was aware of the hostname issue, didn't even know of the static IP bug.

Good to have this fixed, thanks!
